### PR TITLE
The side panel timer begins before the first widget is loaded

### PR DESCRIPTION
**Pull request by bbazyk**
ticket id #178304916
(This branch was created off of chore: WidgetController refactor #252)

### CODE REVIEW

File: app/javascript/components/widget-controller.jsx [hunk 1]
```diff
@@ -33,234 +29,58 @@ const getWidgets = gql`
...
-      );
-    }
-    if (keyCode === downKey) {
-      this.nextWidget();
-    }
-    if (keyCode === upKey) {
-      this.prevWidget();
-    }
-  };
+  const fetchPolicy = newWidgetId === 0 ? 'network-only' : 'cache-first';
```
Comment (sandipa12) [inline app/javascript/components/widget-controller.jsx:35]: what does this line do ?

Comment (bbazyk) [reply app/javascript/components/widget-controller.jsx:35]: I'm curious too - I believe this was a line added in tucker's branch? @tucker-weed

Comment (tucker-weed) [reply app/javascript/components/widget-controller.jsx:35]: The fetch policy is maintained from the previous version of WidgetController - my guess about what is going on here is that the first ever query run by the WidgetController (when `newWidgetId` is 0) should make a network request, whereas all subsequent queries should check if needed query information is in the cache before making a network request. Not 100% sure about this though so if anyone else knows what's really going on here, let us know.

Comment (cbarber) [reply app/javascript/components/widget-controller.jsx:35]: Filling in the missing piece. Its because we prefetch with a direct call to `client.query`. The flow:
* Page load / First request. `<Query`  uses network-only to get the latest data to make sure we're not stale. Updates cache for first widget.
* 1.5 seconds before transition, call `client.query`. This is always a network request. Updates cache for second widget.
* Transition time. Update state to next widget. `<Query` is not `cache-first`. `<Query` has a cache hit from the update from `client.query` 1.5 seconds ago.
* Repeat the previous two steps.

If we only use `network-only` without the prefetch, our transitions are delayed by the server response time. This is compounded by also loading assets after so it was visibly slow.
If we only use `cache-first` our very first request will be our last widget from page close with old time/weather data.

File: app/javascript/components/get-weather-query.jsx [hunk 2]
```diff
@@ -43,49 +43,56 @@ export const WeatherQuery = ({
...
-    variables={{ cityName }}
-  >
-    {({ loading, error, data, subscribeToMore }) => {
-      if (loading) {
-        return <LoadingMessage />;
-      }
-      if (error) {
+  resetTimer,
+}) => {
+  return (
```
Comment (cbarber) [inline app/javascript/components/get-weather-query.jsx:48]: Do we need to switch from the implicit return to an explicit return? If so, lets make that a refactor that precedes the actual change

File: app/javascript/components/get-weather-query.jsx [hunk 3]
```diff
@@ -43,49 +43,56 @@ export const WeatherQuery = ({
...
-  <Query
-    query={getLocationWeather(WeatherFrag, queryName)}
-    variables={{ cityName }}
-  >
-    {({ loading, error, data, subscribeToMore }) => {
-      if (loading) {
-        return <LoadingMessage />;
-      }
-      if (error) {
+  resetTimer,
```
Comment (cbarber) [inline app/javascript/components/get-weather-query.jsx:46]: Maybe a better name? `resetTimer` had me thinking we reset the timer back to 0 when we're completed/errored but I'm guessing this call starts the transition timer.

File: app/javascript/components/widget-display.jsx [hunk 4]
```diff
@@ -72,23 +72,19 @@ export const WidgetDisplay = ({
...
-  }, [current.id]);
+  const resetTimer = () => {
+    dispatch({ type: 'startTimer', durationSeconds: widgetTransitionTime });
+  };
 
   useEffect(() => {
     const timerId = setInterval(() => {
       const { durationRemaining, start } = state;
-      if (isPaused(state)) {
+      if (isPaused(state) || !durationRemaining) {
```
Comment (cbarber) [inline app/javascript/components/widget-display.jsx:82]: What does `!durationremaining` mean here? Could we name/extract that as a selector so its self documenting?

File: app/javascript/components/widgets/twitter/panel.jsx [hunk 5]
```diff
@@ -61,54 +61,16 @@ export const TabBar = () => (
...
-export default () => (
-  <Query query={getMojoTweets}>
+const Twitter = ({ resetTimer }) => (
+  <Query
+    query={getMojoTweets}
+    onCompleted={() => {
+      resetTimer();
+    }}
+    onError={() => {
+      resetTimer();
```
Comment (cbarber) [inline app/javascript/components/widgets/twitter/panel.jsx:71]: Do we need to wrap resetTimer in another function here?

File: app/javascript/components/get-weather-query.jsx (hunk 3 above)
Comment (jennyckaplan) [reply app/javascript/components/get-weather-query.jsx:46]: Maybe back to `startTimer`, since that's the action that this function is dispatching?

**Comment by chewnoill** [comment]
> _:beep: :boop:_ needs a rebase
> 
> :robot:

File: app/javascript/components/widget-display.jsx [hunk 6]
```diff
@@ -72,23 +72,19 @@ export const WidgetDisplay = ({
...
   });
 
-  useEffect(() => {
-    dispatch({
-      type: 'startTimer',
-      durationSeconds: widgetTransitionTime,
-    });
-  }, [current.id]);
+  const resetTimer = () => {
+    dispatch({ type: 'startTimer', durationSeconds: widgetTransitionTime });
```
Comment (chewnoill) [inline app/javascript/components/widget-display.jsx:76]: nit: not really a reset action, since we're not starting on mount anymore.

File: app/javascript/components/widget-display.jsx (hunk 4 above)
Comment (chewnoill) [reply app/javascript/components/widget-display.jsx:82]: 👍 this is a good idea. I think the idea we want is:

```js
function selectIsTimerRunning(state){
  return !isPaused(state) && state.durationRemaining
}
```

**Review by chewnoill — COMMENTED** [review]
Couple nits, nothing major. Kind of sucks that we have to pass these callback functions around so much. Adding Redux is looking better and better.

File: app/javascript/components/widgets/twitter/panel.jsx [hunk 7]
```diff
@@ -61,54 +61,16 @@ export const TabBar = () => (
...
-          <React.Fragment key={tweet.createdAt}>
-            <Tweet tweet={tweet} isPrimary={false} />
-            <TabBar />
-          </React.Fragment>
-        ))}
-      </PreviousWrapper>
-    </PanelWrapper>
-  );
-};
-
```
Comment (tucker-weed) [inline app/javascript/components/widgets/twitter/panel.jsx:82]: Nice, I think combining this Twitter component with its parent component below is much clearer because the parent component isn't named (the parent component was the anonymous function `export default () => ( ...`)

**Comment by bbazyk** [comment]
ready for re-review! @chewnoill @cbarber

File: app/javascript/components/widget-display.jsx [hunk 8]
```diff
@@ -101,6 +103,7 @@ export const LoadingDisplay = ({ cityName, loading }) => (
       <LoadingMessage />
     </FullPanel>
     <SidePanel
+      startTimer={() => {}}
```
Comment (cbarber) [inline app/javascript/components/widget-display.jsx:106]: Why are these noops?

File: app/javascript/components/widgets/numbers/panel.jsx [hunk 9]
```diff
@@ -134,8 +134,17 @@ SubscribedEvents.propTypes = {
...
 
-const Numbers = () => (
-  <Query query={getEventCounts} variables={{ after: getStartOfWeek() }}>
+const Numbers = ({ startTimer }) => (
+  <Query
+    query={getEventCounts}
+    variables={{ after: getStartOfWeek() }}
+    onCompleted={() => {
+      startTimer();
+    }}
```
Comment (cbarber) [inline app/javascript/components/widgets/numbers/panel.jsx:143]: Do we need to wrap startTimer in an anonymous function?

File: app/javascript/components/widgets/twitter/panel.jsx [hunk 10]
```diff
@@ -61,54 +61,8 @@ export const TabBar = () => (
...
-        link: PropTypes.string,
-      }),
-      user: PropTypes.shape({
-        name: PropTypes.string.isRequired,
-        handle: PropTypes.string.isRequired,
-        avatar: PropTypes.string.isRequired,
-      }).isRequired,
-    }),
-  ).isRequired,
-};
```
Comment (cbarber) [inline app/javascript/components/widgets/twitter/panel.jsx:108]: This looks like a refactor in the middle of a feature.
* Why are we consolidating these two components?
* If we need to consolidate the components, we should split them into separate commits.

**Review by cbarber — CHANGES_REQUESTED** [review]

File: app/javascript/components/widgets/twitter/panel.jsx (hunk 10 above)
Comment (bbazyk) [reply app/javascript/components/widgets/twitter/panel.jsx:108]: I've removed all aspects of the refactor now besides references to startTimer, and renaming of the default component - it's in a separate commit from the other changes.

Comment (cbarber) [reply app/javascript/components/widgets/twitter/panel.jsx:108]: If you're undoing changes you no longer, they should be squashed in to the originating commit. Otherwise the edit/undo is just noise.

If you're renaming the default component, that's a refactor. And it should precede your feature commit as a separate unit of work.

File: app/javascript/components/widget-transition-controls.jsx [hunk 11]
```diff
@@ -13,6 +13,8 @@ export const selectShouldRequestPrefetch = state =>
     prefetchDelay;
 
 export const selectIsPaused = state => state.startTimestamp === null;
+export const isStarted = state =>
+  state.startTimestamp !== null && state.msRemainingSinceLastStart !== null;
```
Comment (cbarber) [inline app/javascript/components/widget-transition-controls.jsx:17]: Can you add some unit tests for this?

File: app/javascript/components/get-weather-query.jsx [hunk 12]
```diff
@@ -96,6 +100,11 @@ WeatherQuery.propTypes = {
   }).isRequired,
   queryName: PropTypes.string.isRequired,
   cityName: PropTypes.string.isRequired,
+  startTimer: PropTypes.func,
+};
+
+WeatherQuery.defaultProps = {
+  startTimer: () => {},
```
Comment (cbarber) [inline app/javascript/components/get-weather-query.jsx:107]: IIRC, this noop is used for the weatherCorner. Could we explicitly pass a `startTimer={() => {}}` instead of relying on the default? It makes it seem more intentional imo.

File: app/javascript/components/widget-transition-controls.test.jsx [hunk 13]
```diff
@@ -14,6 +15,12 @@ describe('isPaused', () => {
   });
 });
 
+describe('msRemaining', () => {
+  it('returns true if the milliseconds remaining is any non-null value', () => {
+    expect(isStarted({ msRemainingSinceLastStart: null })).toEqual(false);
+  });
```
Comment (cbarber) [inline app/javascript/components/widget-transition-controls.test.jsx:21]: So we're describing the `isStarted` method. Our `it(...)` doesn't match up with the `toEqual` assertion.

Its fine to squash the unit test into the `isStarted` addition.

**Review by cbarber — COMMENTED** [review]
This is a great improvement. Thank you splitting up the commits so they're easier to understand.

I think there is a little more work left to be done with unit testing `isStarted`. We also talked about adding null checks and testing for the other selectors Tucker wrote to help catch future bugs.

File: app/javascript/components/widget-transition-controls.test.jsx (hunk 13 above)
Comment (cbarber) [reply app/javascript/components/widget-transition-controls.test.jsx:41]: I'd like to see the unit tests in the commit that introduce `isStarted`

File: app/javascript/components/widget-transition-controls.test.jsx [hunk 14]
```diff
@@ -54,6 +81,16 @@ describe('shouldRequestPrefetch', () => {
...
   });
+
+  it('returns true, msRemaining as null will still trigger request prefetch (possible error from editing useEffect execution stop)', () => {
+    expect(
+      selectShouldRequestPrefetch({
+        msRemainingSinceLastStart: null,
+        startTimestamp: Date.now() - 21000,
+        prefetchRequestedYet: false,
+      }),
+    ).toEqual(true);
```
Comment (cbarber) [inline app/javascript/components/widget-transition-controls.test.jsx:92]: I disagree with this assertion. I think `selectShouldRequestPrefetch` should either return false (or throw an exception) if either `msRemainingSinceLastStart` or `startTimestamp` are null. Both are an invalid state for calling this function.

Comment (cbarber) [reply app/javascript/components/widget-transition-controls.test.jsx:92]: I think adding the null checking/throwing and this test should be a separate commit that precedes your feature commit. Something like:
* refactor: rename the default twitter export
* fix: add isStarted check to prevent incorrectly calling `selectShouldRequestPrefetch` with a null value
* fix: disallow calling `selectShouldRequestPrefetch` with null `msRemaining`/`startTimestamp`
* feat: start the sidebar timer after the widget's data is loaded

**Comment by cbarber** [comment]
@bbazyk this isn't passing CI and is not properly rebased (empty commit, tests still split into the wrong commits, fix commit doesn't actually fix the bug). Ping me when you have it fixed.

File: app/javascript/components/widget-transition-controls.jsx [hunk 15]
```diff
@@ -41,8 +41,8 @@ const reducer = (state, action) => {
       };
     case 'startTimer':
       return {
-        msRemainingSinceLastStart: action.msDurationSeconds,
-        startTimestamp: selectIsPaused(state) ? null : now,
+        msRemainingSinceLastStart: action.msDurationRemaining,
```
Comment (jennyckaplan) [inline app/javascript/components/widget-transition-controls.jsx:48]: I'm confused by this commit - what does it mean to change `msDurationSeconds` to `msDurationRemaining`, and why are we getting rid of `selectIsPaused(state)` for determining the `startTimestamp`? and by not dispatching the `startTimer` action within `useEffect` of this component, aren't we making a functionality change and not a refactor?

Comment (bbazyk) [reply app/javascript/components/widget-transition-controls.jsx:48]: The variable name 'milliseconds duration seconds' must just have been a typo in a previous refactor, and I thought msDurationRemaining was more appropriate anyway. For 'selectIsPaused(state) ? null : now', Craig and I decided it's a fair assumption that a widget transition will never occur during a pause, and have to start the timer. I decided on labeling the commit a refactor since the startTimer dispatch wasn't removed from this PR - but instead moved to the final feature commit which creates and passes startTimer.

File: app/javascript/components/widget-transition-controls.jsx [hunk 16]
```diff
@@ -41,8 +41,8 @@ const reducer = (state, action) => {
...
+        startTimestamp: now,
```
Comment (jennyckaplan) [inline app/javascript/components/widget-transition-controls.jsx:49]: Should this be `Date.now()`?

File: app/javascript/components/widget-transition-controls.jsx (hunk 15 above)
Comment (jennyckaplan) [reply app/javascript/components/widget-transition-controls.jsx:48]: Ok, I think this commit would benefit from a more descriptive commit message. Regarding the `selectIsPaused(state)` change, can't we transition widgets using the arrow keys, even when paused? So widget transitions can occur while paused? Would this result in a timer starting on the next widget while paused if transitioning via arrow keys?

Comment (bbazyk) [reply app/javascript/components/widget-transition-controls.jsx:48]: Yes, the timer does start on the next widget, but it is not paused. I've tested transitioning widgets from a paused widget, and couldn't find any other possible errors caused by this action. When helios is paused, and then manually transitioned, the next widget loads just as fine before calling startTimer, and then proceeds unpaused. I guess it depends if we want to consider 'pausing, transitioning, and helios automatically unpausing itself' a bug or a feature?

Comment (bbazyk) [reply app/javascript/components/widget-transition-controls.jsx:48]: I do consider this to be convenient though.

Comment (jennyckaplan) [reply app/javascript/components/widget-transition-controls.jsx:48]: This commit is then changing the behavior (if we're paused, we continue to stay paused until resuming, even if transitioning widgets), so it should be in a different commit if that is your intention in changing that functionality. I personally think we should keep the current functionality and stay paused until we resume.

**Review by bbazyk — COMMENTED** [review]
The setInterval 'refactor-reversion' commit seeks to reverse changes made by an earlier refactor. These changes produced a bug (exposed by implementing startTimer) which could cause widgets to skip over themselves on loading.

**Comment by bbazyk** [comment]
ready for review ! @cbarber

**Review by jennyckaplan — APPROVED** [review]
LGTM

File: app/javascript/components/widget-transition-controls.jsx [hunk 17]
```diff
@@ -75,7 +75,7 @@ const WidgetTransitionControls = ({
   const widgetTransitionTime = current.durationSeconds * 1000;
 
   const [state, dispatch] = useReducer(reducer, {
-    startTimestamp: Date.now(),
+    startTimestamp: null,
```
Comment (planetmcd) [inline app/javascript/components/widget-transition-controls.jsx:80]: What bug is this fixing?

Comment (bbazyk) [reply app/javascript/components/widget-transition-controls.jsx:80]: startTimestamp's value isn't meant to be set in this reducer (and wasn't until a recent refactor PR which accidentally changed it). These changes produced a bug (exposed by implementing startTimer) which could cause widgets to skip over themselves.

Comment (planetmcd) [reply app/javascript/components/widget-transition-controls.jsx:80]: This would be good to include in the commit message.  A format like this:

```
Main Description

Detail for a review that would be helpful for them to understand why you making the changes you are are without them having to run it and recreate.
```

Comment (bbazyk) [reply app/javascript/components/widget-transition-controls.jsx:80]: This does look a lot better. I think I'm going to write a lot more commits like this going forward. Thanks, Mike

**Comment by planetmcd** [comment]
:shipit:
